topics/07_web_ui_testing/main_web_automation_with_playwright.md

```diff
@@ -1,3 +1,488 @@
+    def __init__(self, page: Page):
+        self.page = page
+        # list of selectors
+        self.fullNameField = page.locator("#userName")
+        self.emailField = page.locator("#userEmail")
+        self.currentAddressField = page.locator("#currentAddress")
+        self.permanentAddressField = page.locator("#permanentAddress")
+        self.submitButton = page.locator("#submit")
+        self.pageLoadingIndicator = page.locator(".text-center")
+
+3. Methods of PO
+    # method to iteract with page
+    def navigate(self, url: str):
+        self.page.goto(url)
+        self.currentAddressField.click()
+
+    # method to get information
+    def navigate(self, url: str):
+        self.permanentAddressField.text_content()
+
+    # Assertion methods
+    def check_loading_page(self):
+        return self.fullNameField.text_content() == "John Smith"
+```
+> [!TIP]
+> locator can be organized like 'chain':
+
+```python
+self.access_block = page.locator("//*div[@id='add-visibility-form']")
+self.access_block_select_type = self.access_block.locator("/div[@id='id_requied_action']") 
+instead of "//*div[@id='add-visibility-form']/div[@id='id_requied_action']"
+```
+
+### Assertions
+Playwright has specific assertion: expect(locator).assert_method()
+Assert methods are numerous:
+expect(locator).to_be_hidden()	Element is not visible
+expect(locator).to_be_in_viewport()	Element intersects viewport
+expect(locator).to_be_visible()	Element is visible
+expect(locator).to_contain_text()	Element contains text
+expect(locator).to_have_accessible_description()	Element has a matching accessible description
+expect(locator).to_have_accessible_name()	Element has a matching accessible name
+expect(locator).to_have_attribute()	Element has a DOM attribute
+..
+
+> [!TIP]
+> The main advantage of 'expect' - waiting for page loading with required parameter till framework's timer
+
+#### 'expect' vs default Python 'assert'
+- 'assert' returns Boolean
+- 'expect' returns None
+
+example:
+```python
+    def is_macros_not_displayed_on_page(self, macro_name: str) -> None:
+        expect(self.page.locator(f"//*[text()[contains(.,'{macros_name}')]]")).not_to_be_visible()
+```
+compare with:
+```python
+    def is_macros_not_displayed_on_page(self, macro_name: str) -> bool:
+        return self.page.locator(f"//*[text()[contains(.,'{macros_name}')]]").not_to_be_visible()
+```
+The last one can be used in test in way like:
+
+```python
+assert request_macros_page.is_macro_displayed_on_page(setup_ui_data["macro_name"])
+```
+It more convenient for code but sometimes not convenient in case of delay with page loading
+
+
+
+### Additional features of Playwright vs Pytest
+
+1. Playwright supports all standart types of locators like css or xpath. Additionally it provides the additional functionality like last and first locator:
+
+```python
+page.locator(".classname").last.click()
+```
+2. Ability operate with locator in "#shadow-root"
```

Comment:
Do you have an example of this? Like, an image showing that

Reply:
The example is added.